[mdoc] Preserving remarks created post-factum on import

One of the issues that we are running when we are importing documentation from docXML is that any remarks that were created in the post-production ECMAXML are erased in favor of the imported ones. Here is the scenario.

1. We onboard a new set of DLLs and import Roslyn-generated docXMLs, giving us a new batch of ECMAXML files.
2. Writers edit the ECMAXML remarks to add code examples.
3. A new version of a packages comes out, so we re-run the `mdoc` to import the new information.
4. The import will overwrite the remarks already in ECMAXMLs in favor of those that are in the files already.

This issue was brought up by @mmacy with the Azure Storage libraries, that rely heavily on remarks and user-created samples.

it would seem that this created a regression:

```
2017-10-28T05:11:15.6822016Z Updating: Microsoft.Azure.Documents.SystemFunctions.TypeCheckFunctionsExtensions
2017-10-28T05:11:19.3056944Z ##[error]mdoc: System.Xml.XPath.XPathException: Expression must evaluate to a node-set.
   at MS.Internal.Xml.XPath.XPathParser.ParseNodeTest(AstNode qyInput, AxisType axisType, XPathNodeType nodeType)
   at MS.Internal.Xml.XPath.XPathParser.ParseStep(AstNode qyInput)
   at MS.Internal.Xml.XPath.XPathParser.ParseRelativeLocationPath(AstNode qyInput)
   at MS.Internal.Xml.XPath.XPathParser.ParseLocationPath(AstNode qyInput)
   at MS.Internal.Xml.XPath.XPathParser.ParsePathExpr(AstNode qyInput)
   at MS.Internal.Xml.XPath.XPathParser.ParseUnionExpr(AstNode qyInput)
   at MS.Internal.Xml.XPath.XPathParser.ParseUnaryExpr(AstNode qyInput)
   at MS.Internal.Xml.XPath.XPathParser.ParseMultiplicativeExpr(AstNode qyInput)
   at MS.Internal.Xml.XPath.XPathParser.ParseAdditiveExpr(AstNode qyInput)
   at MS.Internal.Xml.XPath.XPathParser.ParseRelationalExpr(AstNode qyInput)
   at MS.Internal.Xml.XPath.XPathParser.ParseEqualityExpr(AstNode qyInput)
   at MS.Internal.Xml.XPath.XPathParser.ParseAndExpr(AstNode qyInput)
   at MS.Internal.Xml.XPath.XPathParser.ParseOrExpr(AstNode qyInput)
   at MS.Internal.Xml.XPath.XPathParser.ParseExpresion(AstNode qyInput)
   at MS.Internal.Xml.XPath.XPathParser.ParseXPathExpresion(String xpathExpresion)
   at System.Xml.XPath.XPathExpression.Compile(String xpath, IXmlNamespaceResolver nsResolver)
   at System.Xml.XPath.XPathNavigator.Select(String xpath)
   at System.Xml.XmlNode.SelectNodes(String xpath)
   at Mono.Documentation.Updater.MsxdocDocumentationImporter.ImportDocumentation(DocsNodeInfo info)
   at Mono.Documentation.MDocUpdater.MakeDocNode(DocsNodeInfo info, IEnumerable`1 setimporters)
   at Mono.Documentation.MDocUpdater.UpdateMember(DocsNodeInfo info, FrameworkTypeEntry typeEntry)
   at Mono.Documentation.MDocUpdater.DoUpdateType2(String message, XmlDocument basefile, TypeDefinition type, FrameworkTypeEntry typeEntry, String output, Boolean insertSince)
   at Mono.Documentation.MDocUpdater.DoUpdateType(TypeDefinition type, FrameworkTypeEntry typeEntry, String basepath, String dest)
   at Mono.Documentation.MDocUpdater.DoUpdateAssembly(AssemblySet assemblySet, AssemblyDefinition assembly, XmlElement index_types, String source, String dest, HashSet`1 goodfiles)
   at Mono.Documentation.MDocUpdater.DoUpdateAssemblies(String source, String dest)
   at Mono.Documentation.MDocUpdater.Run(IEnumerable`1 args)
   at Mono.Documentation.MDoc.Run(String[] args)
   at Mono.Documentation.MDoc.Main(String[] args)
See `mdoc help' for more information or use --debug to diagnose.
```

@asokolovskiy15 ... see if you can try to recreate the issue in an integration test. Look at calls in `MsxdocDocumentationImporter.ImportDocumentation` that would result in an xpath exception if they don't result in a node set.

--- a/mdoc/Consts.cs
+++ b/mdoc/Consts.cs
@@ -3,7 +3,7 @@ namespace Mono.Documentation
 {
 	public static class Consts
 	{
-		public static string MonoVersion = "5.3.0.0";
+		public static string MonoVersion = "5.3.1.0";
 		public const string DocId = "DocId";
 		public const string VbNet = "VB.NET";
 		public const string DocIdLowCase = "docid";

--- a/mdoc/Mono.Documentation/Updater/MsxdocDocumentationImporter.cs
+++ b/mdoc/Mono.Documentation/Updater/MsxdocDocumentationImporter.cs
@@ -125,14 +125,18 @@ namespace Mono.Documentation.Updater
                         }
                     default:
                         {
-                            bool add = true;
-                            if (child.NodeType == XmlNodeType.Element &&
-                                e.SelectNodes (child.Name).Cast<XmlElement> ().Any (n => n.OuterXml == child.OuterXml))
-                                add = false;
-                            if (add)
+                            var targetNodes = e.ChildNodes.Cast<XmlNode> ()
+                                .Where (n => n.Name == child.Name)
+                                .Select (n => new
+                                {
+                                    Xml = n.OuterXml,
+                                    Overwrite = n.Attributes["overwrite"]
+                                });
+                            string sourceXml = child.OuterXml;
+
+                            if (!targetNodes.Any (n => sourceXml.Equals (n.Xml) || n.Overwrite?.Value == "false"))
                             {
-                                if (child.NodeType == XmlNodeType.Text || e.SelectSingleNode(child.Name) == null)
-                                    MDocUpdater.CopyNode(child, e);
+                                MDocUpdater.CopyNode(child, e);
                             }
                             break;
                         }

--- a/mdoc/mdoc.nuspec
+++ b/mdoc/mdoc.nuspec
@@ -2,7 +2,7 @@
 <package >
   <metadata>
     <id>mdoc</id>
-    <version>5.3.0.0</version>
+    <version>5.3.1.0</version>
     <title>mdoc</title>
     <authors>Joel Martinez</authors>
     <owners>Xamarin</owners>

